src/merkle_tree/constraints.rs

```diff
@@ -101,39 +45,166 @@ where
         let ns = cs.into();
         let cs = ns.cs();
         f().and_then(|val| {
-            let mut path = Vec::new();
-            for &(ref l, ref r) in val.borrow().path.iter() {
-                let l_hash = HGadget::OutputVar::new_variable(
-                    ark_relations::ns!(cs, "l_child"),
-                    || Ok(l),
+            let leaf_sibling = LeafH::OutputVar::new_variable(
+                ark_relations::ns!(cs, "leaf_sibling"),
+                || Ok(val.borrow().leaf_sibling_hash.clone()),
+                mode,
+            )?;
+            let leaf_position_bit = Boolean::new_variable(
+                ark_relations::ns!(cs, "leaf_position_bit"),
+                || Ok(val.borrow().leaf_index & 1 == 1),
+                mode,
+            )?;
+            let pos_list: Vec<_> = val.borrow().position_list().collect();
+            let mut path = Vec::with_capacity(pos_list.len() - 1);
+            for &bit in &pos_list[..pos_list.len() - 1] {
+                // we do not include leaf bit here, so ignore last element returned by pos_list
```

Comment:
You can write this as `let path: Vec<Boolean<_>> = Vec::new_variable(..., || Ok(&pos_list[..(pos_list.len() - 1)]))?;`

Reply:
fixed